exploration/0002-expression-attributes.md

```diff
@@ -0,0 +1,126 @@
+# Expression Attributes
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@eemeli</dd>
+		<dt>First proposed</dt>
+		<dd>2023-08-27</dd>
+		<dt>Pull Request</dt>
+		<dd>#000</dd>
+	</dl>
+</details>
+
+## Objective
+
+Define how attributes may be attached to expressions.
+
+## Background
+
+Function options may influence the resolution, selection, and formatting of annotated expressions.
+These provide a great solution for options like `minFractionDigits`, `dateStyle`,
+or other similar factors that influence the formatted result.
+
+However, this single bag of options is not appropriate in all cases,
+in particular for attributes that pertain to the expression as a selector or a placeholder.
+For example, many of the [XLIFF 2 inline element] attributes don't really make sense as function options.
+
+[XLIFF 2 inline element]: http://docs.oasis-open.org/xliff/xliff-core/v2.1/os/xliff-core-v2.1-os.html#inlineelements
+
+## Use-Cases
+
+At least the following expression attributes should be considered:
+
+- Attributes with a formatting runtime impact:
+
+  - `fallback` — A value to use instead of the default fallback,
+    should the expression's primary formatting fail in some way.
+  - `locale` — An override for the locale used to format the expression.
+    Should be expressed as a non-empty sequence of BCP 47 language codes.
+  - `dir` — An override for the LTR/RTL/auto directionality of the expression.
+
+- Attributes relevant for translators, tools, and other message operations,
+  but with no runtime impact:
+
+  - `example` — A literal value representing
+    what the expression's formatted value will look like.
+  - `note` — A comment on the expression for translators.
+  - `translate` — A boolean `yes`/`no` indicator communicating to translators
+    whether the expression should or should not be localised.
+  - `canCopy`, `canDelete`, `canOverlap`, `canReorder`, etc. — Flags supported by
+    XLIFF 2 inline elements
+
+## Requirements
+
+Attributes should be assignable to un-annotated expressions.
+
+Attributes should be distinct from function options.
+
+Common attributes should be definable by the specification.
+
+Attributes should be user-definable.
+
+Attributes with a runtime impact should not be considered comments.
```

Comment:
This is intended as a reason why e.g. semantic comments are not a suitable solution for all attributes.

Would this clarify the intended meaning?
```suggestion
Attributes with a runtime impact must not be defined as comments.
```

Reply:
I see: I'm thinking about this requirement from a different point of view. I think the change doesn't communicate this quite yet. Perhaps:

```suggestion
In order to allow attributes that have runtime impact, they must appear in the expression and not as separate values, such as comments.
```

I would also probably consider adding:

```suggestion
Attributes that affect the localization of a pattern must be part of the expression, rather than separate, to ensure that they are not lost during translation.
```

Reply:
How about this?

```suggestion
Some attributes may have an effect on the formatting of an expression.
These cannot be defined within comments either within or outside a message.
```